### PR TITLE
Added in-process embedding models

### PR DESCRIPTION
**Pull request by ghost**
- all-minilm-l6-v2
- all-minilm-l6-v2-q
- e5-small-v2
- e5-small-v2-q

The idea is to give users an option to embed documents/texts in the same Java process without any external dependencies.
ONNX Runtime is used to run models inside JVM.
Each model resides in it's own maven module (inside the jar).

### CODE REVIEW

File: langchain4j/src/main/java/dev/langchain4j/model/inprocess/InProcessEmbeddingModel.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package dev.langchain4j.model.inprocess;
+
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+
+import java.util.List;
+
+import static java.lang.String.format;
+
```
Comment (kuraleta) [inline langchain4j/src/main/java/dev/langchain4j/model/inprocess/InProcessEmbeddingModel.java:10]: please add documentation

File: langchain4j/src/main/java/dev/langchain4j/model/inprocess/InProcessEmbeddingModelType.java
Comment (kuraleta) [inline langchain4j/src/main/java/dev/langchain4j/model/inprocess/InProcessEmbeddingModelType.java:1]: I like links with details!

File: langchain4j-embeddings-all-minilm-l6-v2-q/src/main/java/dev/langchain4j/model/embedding/all_minilm_l6_v2_q_EmbeddingModel.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package dev.langchain4j.model.embedding;
+
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+public class all_minilm_l6_v2_q_EmbeddingModel implements EmbeddingModel {
```
Comment (kuraleta) [inline langchain4j-embeddings-all-minilm-l6-v2-q/src/main/java/dev/langchain4j/model/embedding/all_minilm_l6_v2_q_EmbeddingModel.java]: please name the embedding model classes as an enum

File: langchain4j-embeddings-all-minilm-l6-v2-q/src/main/java/dev/langchain4j/model/embedding/all_minilm_l6_v2_q_EmbeddingModel.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            "/all-minilm-l6-v2-q.onnx",
+            "/vocab.txt"
+    );
+
+    @Override
+    public List<Embedding> embedAll(List<TextSegment> segments) {
+        return segments.stream()
+                .map(segment -> {
+                    float[] vector = model.embed(segment.text());
+                    return Embedding.from(vector);
```
Comment (Shastick) [inline langchain4j-embeddings-all-minilm-l6-v2-q/src/main/java/dev/langchain4j/model/embedding/all_minilm_l6_v2_q_EmbeddingModel.java]: Absolute nitpick: we could inline `vector` and make the whole `.map` statement simpler?

Additionally, `EmbeddingModel` could have a default implementation for `embedAll` and obtain `model` through a getter? 

edit: though I'm not entirely sure if that will be helpful...

**Comment by Shastick** [comment]
If I got it right, we're essentially adding these models to this repository?

If we're going to do this for more models in the future, might it be better to download the models during the build/release process? 

Another question I wonder about, if these models are a bit heaver: is sonatype fine with having libs wrap things that aren't really jars? And should we or could we release these on a slower cadence so as not to have a lot of bigger duplicated jars?

In an ideal world, there would be a way to download said models from eg hugging face when projects that depend on this are building their code, but I'm not sure how possible that is.

File: langchain4j-embeddings-all-minilm-l6-v2-q/src/main/java/dev/langchain4j/model/embedding/all_minilm_l6_v2_q_EmbeddingModel.java (hunk 3 above)
Comment (langchain4j) [reply langchain4j-embeddings-all-minilm-l6-v2-q/src/main/java/dev/langchain4j/model/embedding/all_minilm_l6_v2_q_EmbeddingModel.java]: yes, makes sense to extract this logic to some base class, as it is now duplicated in each model

**Comment by langchain4j** [comment]
> If I got it right, we're essentially adding these models to this repository?

Yes, I am looking into git LFS for this, it seems doable on github.

> If we're going to do this for more models in the future, might it be better to download the models during the build/release process?

I was considering downloading during build but I am not sure it is a good idea to rely on some external service where people can modify/delete models and that can be down at some point.

> Another question I wonder about, if these models are a bit heaver: is sonatype fine with having libs wrap things that aren't really jars?

As far as I read sonatype does not have a limit on the file size.

> And should we or could we release these on a slower cadence so as not to have a lot of bigger duplicated jars?

Not sure I understand this question, could you please elaborate?

> In an ideal world, there would be a way to download said models from eg hugging face when projects that depend on this are building their code, but I'm not sure how possible that is.

Right now models are downloaded from maven during build time (models are inside optional modules like langchain4j-embeddings-e5-small-v2). This is basically the same, but we have control over models and make sure they will not change because someone decided to update their model on hugging face.

**Comment by Shastick** [comment]
> > And should we or could we release these on a slower cadence so as not to have a lot of bigger duplicated jars?
> 
> Not sure I understand this question, could you please elaborate?

We will probably be releasing this library pretty often in the near future, I was wondering if it will be wise to re-release all the models each time if they don't change, but I'm probably overthinking this, so nevermind.



> >    If I got it right, we're essentially adding these models to this repository?
> Yes, I am looking into git LFS for this, it seems doable on github.

How big are the models, typically? if it's a few hundreds of kbs to maybe a few megabytes, it's probably fine to add them to git directly (as long as the models don't change much :))

Otherwise: there is the following warning in the build:


```
 Warning:  Checksum validation failed, no checksums available from central for https://repo.maven.apache.org/maven2/com/microsoft/onnxruntime/onnxruntime/1.15.1/onnxruntime-1.15.1.jar
```

Is this to be expected?

**Comment by deep-learning-dynamo** [comment]
Would be cool to not re-release them every time. Yesterday it was a pain to
release 4 of them, sonatype was constantly timeouting. On the other side it
would be nice to have the same version consistent over all the modules...

They are 20-100MB now each, but can be bigger. It is not clear yet if users
will find it practical to import even bigger models...probably not.

**Comment by langchain4j** [comment]
I have added models ( .onnx files ) into LFS, but it seems that they are not found during the build. Tests in this branch are failing and it seems that .onnx files cannot be found or loaded (maybe due to those links from git to git LFS)... I did not look deep into it yet. Regarding onnxruntime warning, not sure what's the problem. Does not looks critical. This lib is used to run the models, so we can't drop it or find an alternative...

**Comment by langchain4j** [comment]
@Shastick I had to add this part in order to make it work:
```
      - name: Pull LFS files
        run: git lfs pull
```
Please feel free to change/improve.

**Review by kuraleta — APPROVED** [review]

**Comment by langchain4j** [comment]
@Shastick it seems that free LFS traffic quota (1GB/month) is used every time the build is running and it downloads all .onnx models. Can we make it run those tests only if there were changes in tose specific modules?

**Comment by Shastick** [comment]
@langchain4j adding such rules is possible albeit can become annoying. Let's try like it is now to see how we can keep things simple.

If you want to add a bigger collection of models available in-process there is also the option of having them in a separate repository.

**Comment by langchain4j** [comment]
@Shastick good idea regarding separate repository. These models will rarely change, so it probably makes no sense to keep them in the main repo.

**Comment by langchain4j** [comment]
@Shastick BTW I have removed "git lfs pull" and disabled tests cause we were out of free quota...

**Comment by Shastick** [comment]
Do you know what ate up the quotas? LFS or the test minutes?

**Comment by langchain4j** [comment]
@Shastick I guess .onnx files from LFS. It looks like they were downloaded each time the build was running..

**Comment by Shastick** [comment]
Then I guess a separate repo is warranted. I might be able to give it a look this week-end.

**Pull request merged**